### PR TITLE
HDDS-9441. Support Recon om snapshot full update interval delay

### PR DESCRIPTION
**Pull request by aierate**
## Background

There are two methods for retrieving om roskDB snapshots from om: incremental update and full update. Incremental update may result in inaccurate update data due to cluster operations and other reasons, resulting in inaccurate record statistics. Therefore, it is necessary to add a timed full update function to update the om roskDB snapshot in a timely and full manner, ensuring the accuracy of recon statistical information. The following is the test report.

## Configuration

Enable zone recon to support the XML configuration items recon. om. snapshot. full. update. interval. enable and recon. om. snapshot. full. update. interval. delay. Set the time interval for recon to schedule and fully obtain om roskDB snapshots. The configuration is as follows:

```
<property>
 <name>recon.om.snapshot.full.update.interval.delay</name>
  <value>5m</value>
  <tag>OZONE, RECON</tag>
  <description>
    Recon regularly fetches ozone manager rocks DB from ozone manager and updates recon rocks DB information
  </description>
</property>

<property>
  <name>recon.om.snapshot.full.update.interval.enable</name>
  <value>true</value>
  <tag>OZONE, RECON</tag>
  <description>
    Whether to enable recon to fetch ozone manager rocks DB from ozone manager regularly
  </description>
</property>
```

## Test result
**2023-10-12 12:01:29,922**
```
org.apache.hadoop.ozone.recon.spi.impl.OzoneManagerServiceProviderImpl: Syncing data from Ozone Manager.
2023-10-12 12:01:29,922 [pool-27-thread-1] INFO org.apache.hadoop.ozone.recon.spi.impl.OzoneManagerServiceProviderImpl: Obtaining full snapshot from Ozone Manager
2023-10-12 12:01:33,729 [pool-27-thread-1] INFO org.apache.hadoop.ozone.recon.spi.impl.OzoneManagerServiceProviderImpl: Got new checkpoint from OM : /home/hadoop-addfullUpdateXML-03/ozonedata/omdbdir/om.snapshot.db_1697083289922
2023-10-12 12:01:33,734 [pool-27-thread-1] INFO org.apache.hadoop.ozone.om.helpers.OmKeyInfo: OmKeyInfo.getCodec ignorePipeline = true
2023-10-12 12:01:34,149 [pool-27-thread-1] INFO org.apache.hadoop.ozone.recon.recovery.ReconOmMetadataManagerImpl: Created OM DB handle from snapshot at /home/hadoop-addfullUpdateXML-03/ozonedata/omdbdir/om.snapshot.db_1697083289922.
```

**2023-10-12 12:06:35,387**
```
2023-10-12 12:06:35,386 [pool-27-thread-1] INFO org.apache.hadoop.ozone.recon.spi.impl.OzoneManagerServiceProviderImpl: Syncing data from Ozone Manager.
2023-10-12 12:06:35,387 [pool-27-thread-1] INFO org.apache.hadoop.ozone.recon.spi.impl.OzoneManagerServiceProviderImpl: Obtaining full snapshot from Ozone Manager
2023-10-12 12:06:35,449 [pool-27-thread-1] INFO org.apache.hadoop.ozone.recon.spi.impl.OzoneManagerServiceProviderImpl: Got new checkpoint from OM : /home/hadoop-addfullUpdateXML-03/ozonedata/omdbdir/om.snapshot.db_1697083595387
2023-10-12 12:06:35,450 [pool-27-thread-1] INFO org.apache.hadoop.ozone.recon.recovery.ReconOmMetadataManagerImpl: Cleaning up old OM snapshot db at /home/hadoop-addfullUpdateXML-03/ozonedata/omdbdir/om.snapshot.db_1697083289922.
```

### CODE REVIEW

**Comment by adoroszlai** [comment]
@aierate HDDS-9440 is unrelated.  Is there an existing Jira issue for this change?  If not, please create one.   Please update PR title with the Jira issue ID.

**Comment by aierate** [comment]
[HDDS-9441](https://issues.apache.org/jira/browse/HDDS-9441) This is a Jira issue, and the PR title has been updated. Thank you for your review.

**Comment by adoroszlai** [comment]
@aierate please enable the `build-branch` workflow [in your fork](https://github.com/aierate/ozone/actions/workflows/post-commit.yml) to get initial CI run

File: .github/workflows/docker-image.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+jobs:
+
+  build:
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v3
+    - name: Build the Docker image
+      run: docker build . --file Dockerfile --tag my-image-name:$(date +%s)
```
Comment (adoroszlai) [inline .github/workflows/docker-image.yml:18]: This workflow file seems to be completely unrelated to the PR.

**Comment by aierate** [comment]
Sorry, there are some Checkstyle issues in my code. I will fix them later and resubmit.

**Comment by aierate** [comment]
I need to learn how to use a CI run in my fork on github.

**Comment by dombizita** [comment]
Hi @aierate, thanks for filing this jira and working on it!

> Incremental update may result in inaccurate update data due to cluster operations and other reasons, resulting in inaccurate record statistics.

Did you encounter something like this? Could you detail this a bit more?

A full update in Recon in a big cluster can take a long time as it is going through the whole OM Metadata RocksDB, so we need to be cautious when we want to trigger that.

**Comment by aierate** [comment]
@dombizita Yes, I encountered this problem during the monitoring of Recon. I have about 100 DataNodes here, and the cluster is running normally, and the incremental acquisition of OM snapshots in Recon is also normal, but the data in the OM snapshots in Recon is different from the content of OM snapshots. Therefore, I think there are some unknown issues with the incremental acquisition of OM snapshots in Recon, and Recon needs to update the OM snapshots in full regularly to ensure accuracy.

**Comment by dombizita** [comment]
> the incremental acquisition of OM snapshots in Recon is also normal, but the data in the OM snapshots in Recon is different from the content of OM snapshots. Therefore, I think there are some unknown issues with the incremental acquisition of OM snapshots in Recon

It'd be better to find out what could cause this issue and fix that, I know that it's hard to follow up on these, but ideally we shouldn't have these problems. Could you look into the logs and maybe share the differences you see? I still think that triggering a full update in Recon is expensive and in large clusters it can take a lot of time to process all the events that happened since startup.

Sometime back we added a button to trigger a delta update in Recon (#3821). We talked about having a similar button for full update but I think because of the above mentioned reasons we postponed it. 

It'd be great to have the opinion of others, I'll tag some people as reviewer. If we go with this way I'd suggest to have it turned off by default for sure.

**Comment by ArafatKhan2198** [comment]
Hi aierate 
I completely agree with what @devmadhuu and @dombizita have been saying. Investigating the root cause of the discrepancies between Recon and OM data is essential for maintaining data accuracy. While a full update button could be useful, we should be cautious about its impact, especially in large clusters.

@aierate, could you provide more details about the specific data you were comparing between the OM snapshot and Recon snapshot? We have received a few user complaints regarding inaccuracies in namespace summary creation. I can assure you that these issues are not related to snapshot inconsistencies.

**Comment by aierate** [comment]
Thank you very much, everyone.@dombizita @devmadhuu @ArafatKhan2198 
Recently, due to some reasons（https://github.com/apache/ozone/discussions/5494）, I have stopped Recon and cannot provide detailed information about it. But I think what you said is correct, and once again, thank you very much.
